### PR TITLE
DBZ-4401 Support partition key row deletion and range tombstone row types

### PR DESCRIPTION
**Pull request by smiklosovic**
Hi @gunnarmorling , @bingqinzhou , @ahmedjami 

This patch adds support for partition key row deletion and range tombstones.

For range tombstones, it is introducing new struct called "bounds" with "start" and "end" fields.

This is the implementation for Cassandra 4 only for now. If we agree with the approach, I will try to introduce this to Cassandra 3 too.

There are added tests for each scenario for Cassandra 4 implementation.

Please tell me what you think about this.

### CODE REVIEW

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogReadHandlerImpl.java
Comment (smiklosovic) [inline cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogReadHandlerImpl.java:291]: Previous code had it wrong. For Cassandra 4, we can not cast `RangeTombstoneBoundMarker` to `Row`. In Cassandra 4, `Row` implements `Unfiltered` and `RangeTombstoneBoundMarker` implements `Unfiltered`, but `RangeTombstoneBoundMarker` does not extend nor implement `Row` so we do not have any information about values but the primary key in `PartitionUpdate`.

**Comment by ahmedjami** [comment]
Hi,
Thanks a lot @smiklosovic for this new PR :) 
I'm back and I'm planning to test the new feature soon.
Cheers,
Ahmed

**Comment by smiklosovic** [comment]
hey @ahmedjami wait a day or two and I ll update this stuff on top of the current main branch, this one is little bit behind.

**Comment by smiklosovic** [comment]
@ahmedjami  it is aligned, you can try now.

**Comment by jpechane** [comment]
@vjuranek Could you please take a look?
@ahmedjami I'd be happy to hear about your feedback.

**Comment by ahmedjami** [comment]
Hey!

> @ahmedjami I'd be happy to hear about your feedback.
Yes ;)

File: cassandra-4/src/test/java/io/debezium/connector/cassandra/RangeTombstoneCommitLogProcessingTest.java
Comment (vjuranek) [inline cassandra-4/src/test/java/io/debezium/connector/cassandra/RangeTombstoneCommitLogProcessingTest.java:68]: would be good to test also `start` and `end` of the range delete

File: cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogReadHandlerImpl.java
Comment (vjuranek) [inline cassandra-4/src/main/java/io/debezium/connector/cassandra/Cassandra4CommitLogReadHandlerImpl.java:523]: Would be good to move these (unrelated) cleanup change into separate commit. It makes the review more easy and also backports are much more easy to do when there's a need to backport a commit. Not anything which needs to be done now, just FIY - a suggestion for your next PRs to come.

**Review by vjuranek — COMMENTED** [review]
Overall LGTM, just one minor suggestion for improving the test. But given my very limited knowledge of Cassandra, it would be nice to have more reviews:-)

**Comment by ahmedjami** [comment]
@smiklosovic 

The new implementation works well for _partition key type row deletion_  when I perform a delete query with only the partition key.

However, the connector fails when I try to perform a delete with a range tombstone. Here is the error :

```
INFO  || LogicalCommitLog{commitLogPosition=CommitLogPosition(segmentId=1651752179377, position=0), synced=1048496, completed=true, log=/var/lib/cassandra/cdc_raw_directory/CommitLog-7-1651752179377.log, index=/var/lib/cassandra/cdc_raw_directory/CommitLog-7-1651752179377_cdc.idx, commitLogSegmentId=1651752179377} [io.debezium.connector.cassandra.Cassandra4CommitLogProcessor$CommitLogProcessingCallable]

2022-05-05 12:37:31,119 ERROR || Processing of event Record{source={cluster=cassandra-cluster, keyspace=keyring_prod, file=CommitLog-7-1651752179377.log, connector=cassandra, pos=514545, ts_micro=1651754179795815, version=${project.version}, snapshot=false, table=cred}, after={s={name=s, value=100,3QxeOYuvcrqVcM5SlIcBHR6RGjKRn7BLkIb_rThr, deletionTs=null, type=PARTITION}}, keySchema=Schema{io.debezium.connector.cassandra.test_prefix.keyring_prod.cred.Key:STRUCT}, valueSchema=Schema{io.debezium.connector.cassandra.test_prefix.keyring_prod.cred.Envelope:STRUCT}, op=r, ts=1651754249716} was errorneous: {} [io.debezium.connector.cassandra.QueueProcessor]

io.debezium.DebeziumException: Failed to send record Record{source={cluster=cassandra-cluster, keyspace=keyring_prod, file=CommitLog-7-1651752179377.log, connector=cassandra, pos=514545, ts_micro=1651754179795815, version=${project.version}, snapshot=false, table=cred}, after={s={name=s, value=100,3QxeOYuvcrqVcM5SlIcBHR6RGjKRn7BLkIb_rThr, deletionTs=null, type=PARTITION}}, keySchema=Schema{io.debezium.connector.cassandra.test_prefix.keyring_prod.cred.Key:STRUCT}, valueSchema=Schema{io.debezium.connector.cassandra.test_prefix.keyring_prod.cred.Envelope:STRUCT}, op=r, ts=1651754249716}

    at io.debezium.connector.cassandra.KafkaRecordEmitter.emit(KafkaRecordEmitter.java:71)

    at io.debezium.connector.cassandra.QueueProcessor.processEvent(QueueProcessor.java:114)

    at io.debezium.connector.cassandra.QueueProcessor.process(QueueProcessor.java:72)

    at io.debezium.connector.cassandra.AbstractProcessor.start(AbstractProcessor.java:63)

    at io.debezium.connector.cassandra.CassandraConnectorTaskTemplate$ProcessorGroup.lambda$start$0(CassandraConnectorTaskTemplate.java:231)

    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)

    at java.base/java.util.concurrent.FutureTask.run(Unknown Source)

    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)

    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)

    at java.base/java.lang.Thread.run(Unknown Source)

Caused by: org.apache.kafka.connect.errors.DataException: start is not a valid field name

    at org.apache.kafka.connect.data.Struct.lookupField(Struct.java:254)

    at org.apache.kafka.connect.data.Struct.put(Struct.java:202)

    at io.debezium.connector.cassandra.RowData.record(RowData.java:71)

    at io.debezium.connector.cassandra.Record.buildValue(Record.java:86)

    at io.debezium.connector.cassandra.KafkaRecordEmitter.toProducerRecord(KafkaRecordEmitter.java:81)

    at io.debezium.connector.cassandra.KafkaRecordEmitter.emit(KafkaRecordEmitter.java:62)

    ... 9 common frames omitted


```
Could you take a look please ?

**Comment by smiklosovic** [comment]
Interesting. Yes, I will take a look. Not sure what the bug is though.

**Comment by smiklosovic** [comment]
@ahmedjami can you play with `source.struct.version` property?

**Comment by ahmedjami** [comment]
Here is the schema conf used by Debezium:

```
key.converter.schemas.enable=false
value.converter.schemas.enable=false

key.converter=org.apache.kafka.connect.json.JsonConverter
value.converter=org.apache.kafka.connect.json.JsonConverter
```
But, I don't set the schema registry URL (`key.converter.schema.registry.url`) so it shouldn't be used in my case.

**Comment by smiklosovic** [comment]
ok, I need to test it more.

**Comment by smiklosovic** [comment]
I am on it, digging the rabbit hole. The current implementation is not too much user friendly when it comes to testing internal stuff, it works nicely when we take a look at that from perspective, that works nicely, but as soon as I want to extract and test only some isolated part of the logic, it is not so straightforward. Hence I am spending some time on testability in general.

**Comment by smiklosovic** [comment]
@ahmedjami I think I fixed it, would you mind to take a look again? I am not sure if the format I put these ranges in will be ok for you, that is why I need broader consensus on this. You will likely see it in the message sent to Kafka.

I reworked tests in such a way that it will try to construct the message as it would be sent to Kafka.

I will copy the logic done in cassandra-4 for range tombstones to cassandra-3 once we are ok with this approach.

**Comment by ahmedjami** [comment]
Hi @smiklosovic 
Yes of course :) Will keep you updated.

**Comment by smiklosovic** [comment]
Hey, any progress here?

**Comment by ahmedjami** [comment]
Hey @smiklosovic 
Sorry for the delay, we have tested the fix and I don't show the error :) 
The format looks good to me but I need to discuss and share it with our dev team. It should be confirmed this week.

Thanks,

**Comment by ahmedjami** [comment]
Hey,

@smiklosovic , the message format is OK and all tests are green :) Thanks a lot for this excellent work.

@vjuranek @jpechane PR can be merged if you don't have any other comments :)

Thanks,

**Review by vjuranek — APPROVED** [review]

**Comment by vjuranek** [comment]
Merged, thanks @smiklosovic ! And thanks @ahmedjami for testing!

**Comment by smiklosovic** [comment]
@vjuranek this is not all! :D I need to backport it to cassandra 3 as well, what I did was for Cassandra 4 only. If you go through the comments you will see it what I did. This is misunderstanding. 

So I ll prepare it for cassanra 3 too ok?

**Comment by ahmedjami** [comment]
oh my bad !
I should also not have asked for it to be merged... You will reopen it or create a new ticket for that?

**Comment by smiklosovic** [comment]
I will prepare next PR, it can go under same JIRA ticket.

**Comment by vjuranek** [comment]
@smiklosovic yes, I understand that. Is there any need to do it in the same PR? My understanding is that your approach is fine, @ahmedjami tested it and it works, so you can go ahead and implement it aslo for Cassandra 3. Am I missing something?

**Comment by smiklosovic** [comment]
@vjuranek that is great you get that, I was thinking that after he confirms I will just extend the same approach to Cassandra 3 and we just commit it all at once so we do not need to merge it twice. But if you do not mind me to go and create another PR, be my guest.

**Comment by vjuranek** [comment]
@smiklosovic definitely no problem with another PR on my side. Actually I'd prefer it as going through huge PR is error prone.

**Comment by vjuranek** [comment]
One more thing: after merging this PR, Cassandra 4 build stated to [fail](https://github.com/debezium/debezium-connector-cassandra/actions/runs/2344339482) although before merged [it was passing](https://github.com/debezium/debezium-connector-cassandra/actions/runs/2299826234). Could you please take a look @smiklosovic and eventually fix it to unbreak the build? Thanks!

**Comment by vjuranek** [comment]
@smiklosovic I create PR #56 to fix it - fixes it for me at least locally.

**Pull request merged**
